ledger-service/http-json/src/main/scala/com/digitalasset/http/ContractsFetch.scala

```diff
@@ -289,85 +291,112 @@ private class ContractsFetch(
+                (stepsAndOffset.out0, stepsAndOffset.out1)
+
+              case (AbsoluteBookmark(_), _) | (LedgerBegin, true) =>
+                val stepsAndOffset = builder add transactionsFollowingBoundary(txnK)
+                stepsAndOffset.in <~ Source.single(startOffset)
+                (
+                  (stepsAndOffset: FanOutShape2[_, ContractStreamStep.LAV1, _]).out0,
+                  stepsAndOffset.out1,
+                )
+            }
 
-          val transactInsertsDeletes = Flow
-            .fromFunction(
-              jsonifyInsertDeleteStep(
-                (_: InsertDeleteStep[Any, lav1.event.CreatedEvent]),
-                templateId,
+            val transactInsertsDeletes = Flow
+              .fromFunction(
+                jsonifyInsertDeleteStep(
+                  (_: InsertDeleteStep[Any, lav1.event.CreatedEvent]),
+                  templateId,
+                )
               )
-            )
-            .via(if (sjd.q.queries.allowDamlTransactionBatching) conflation else Flow.apply)
-            .map(insertAndDelete)
+              .via(if (sjd.q.queries.allowDamlTransactionBatching) conflation else Flow.apply)
+              .map(insertAndDelete)
 
-          idses.map(_.toInsertDelete) ~> transactInsertsDeletes ~> acsSink
-          lastOff ~> offsetSink
+            idses.map(_.toInsertDelete) ~> transactInsertsDeletes ~> acsSink
+            lastOff ~> offsetSink
 
-          ClosedShape
-        }
-      )
+            ClosedShape
+          }
+        )
 
-      val (acsQueue, lastOffsetFuture) = graph.run()
+        val (acsQueue, lastOffsetFuture) = graph.run()
+
+        for {
+          _ <- sinkCioSequence_(acsQueue)
+          offset0 <- connectionIOFuture(lastOffsetFuture)
+          offsetOrError <- offset0 max AbsoluteBookmark(absEnd.toDomain) match {
+            case ab @ AbsoluteBookmark(newOffset) =>
+              ContractDao
+                .updateOffset(parties, templateId, newOffset, offsets)
+                .map(_ => ab)
+            case LedgerBegin =>
+              fconn.pure(LedgerBegin)
+          }
+        } yield offsetOrError
+      }
+  }
 
-      for {
-        _ <- sinkCioSequence_(acsQueue)
-        offset0 <- connectionIOFuture(lastOffsetFuture)
-        offsetOrError <- offset0 max AbsoluteBookmark(absEnd.toDomain) match {
-          case ab @ AbsoluteBookmark(newOffset) =>
-            ContractDao
-              .updateOffset(parties, templateId, newOffset, offsets)
-              .map(_ => ab)
-          case LedgerBegin =>
-            fconn.pure(LedgerBegin)
+  private def debugLogAction[T, C](
+      actionDescription: String
+  )(block: => T)(implicit lc: LoggingContextOf[C]): T = {
+    if (logger.debug.isEnabled) {
+      val startTime = System.nanoTime()
+      logger.debug(s"Starting $actionDescription")
+      val result =
+        try {
+          block
+        } catch {
+          case e: Exception =>
+            logger.debug(
```

Comment:
```suggestion
            logger.error(
```
Realised this should probably be `error`

Reply:
I thought about it too, but then the log will be only printed with debug level (cos of the if block `logger.debug.isEnabled`, this is done for performance reasons btw) ...... options here are to rely in the logs that are on top of the `fetchAndPersist` which is the function that call this cache refresh (the message of course is more general), or to don't care about the performance and just don't check the log level that is running atm of the call

Reply:
or yeah just change it to error, but we need to be aware that these will be generated when the log level is on debug mode

Reply:
probably the later is the best accepting the suggestion... cheers ))